datafusion/physical-plan/src/joins/nested_loop_join.rs

```diff
@@ -560,91 +729,147 @@ impl NestedLoopJoinStream {
+                            }
+                        } else {
                             self.is_exhausted = true;
-                            return None;
+                            continue;
+                        }
+                    }
+                    Poll::Pending => {
+                        return match self.output_buffer.flush() {
+                            Ok(Some(batch)) => {
+                                // If there was anything in the output buffer flush it
+                                // so that it can be processed.
+                                self.join_metrics.output_batches.add(1);
+                                self.join_metrics.output_rows.add(batch.num_rows());
+                                Poll::Ready(Some(Ok(batch)))
+                            }
+                            Ok(None) => Poll::Pending,
+                            Err(err) => Poll::Ready(Some(Err(err))),
                         };
+                    }
+                }
+            }
 
-                        // Only setting up timer, input is exhausted
-                        let timer = self.join_metrics.join_time.timer();
-                        // use the global left bitmap to produce the left indices and right indices
-                        let (left_side, right_side) =
-                            get_final_indices_from_shared_bitmap(
-                                visited_left_side,
-                                self.join_type,
-                            );
-                        let empty_right_batch =
-                            RecordBatch::new_empty(self.outer_table.schema());
-                        // use the left and right indices to produce the batch result
-                        let result = build_batch_from_indices(
-                            &self.schema,
-                            left_data.batch(),
-                            &empty_right_batch,
-                            &left_side,
-                            &right_side,
-                            &self.column_indices,
-                            JoinSide::Left,
-                        );
-                        self.is_exhausted = true;
-
-                        // Recording time & updating output metrics
-                        if let Ok(batch) = &result {
-                            timer.done();
-                            self.join_metrics.output_batches.add(1);
-                            self.join_metrics.output_rows.add(batch.num_rows());
-                        }
+            debug_assert!(self.outer_record_batch.is_some());
+            let right_batch = self.outer_record_batch.as_ref().unwrap();
+            let num_rows = match (self.join_type, left_data.batch().num_rows()) {
+                // An inner join will only produce 1 output row per input row.
+                (JoinType::Inner, _) | (_, 0) => self.output_buffer.needed_rows(),
```

Comment:
Isn't it possible for an Inner Join produce multiple output rows per input row? I believe the "1 output row per input row" statement only holds true for equijoins. NestedLoopJoin works on non equijoins by definition.

https://github.com/apache/datafusion/blob/9b4f90ad1eefabdc0d5bbbfd99e58765b041bb77/datafusion/physical-plan/src/joins/nested_loop_join.rs#L103-L104

For example, for `left=[1, 2]` and `right=[1, 2, 3]` with the `ON` clause `left<>right`, it produces `[(1, 2), (2, 1), (1, 3), (2, 3)]`. The row `3` from the right side produces 2 rows.

It seems impossible to predict the number of output rows without running the join.

Reply:
You are right of course, I'm not sure why I got it in my head that they would be 1-1. I'll fix that.